frame/support/procedural/src/benchmark.rs

```diff
@@ -0,0 +1,786 @@
+						::decode(&mut &__benchmarked_call_encoded[..])
+						.expect("call is encoded above, encoding must be correct");
+					let __origin = #origin.into();
+					<Call<#generics> as #traits::UnfilteredDispatchable>::dispatch_bypass_filter(
+						__call_decoded,
+						__origin,
+					)?;
+				},
+			)
+		},
+		BenchmarkCallDef::Block(block) => (quote!(), quote!(#block)),
+	};
+
+	// generate final quoted tokens
+	let res = quote! {
+		// compile-time assertions that each referenced param type implements ParamRange
+		#(
+			#home::assert_impl_all!(#param_types: #home::ParamRange);
+		)*
+
+		#[allow(non_camel_case_types)]
+		struct #name;
+
+		#[allow(unused_variables)]
+		impl<#full_generics> #krate::BenchmarkingSetup<#generics>
+		for #name where #where_clause {
+			fn components(&self) -> #krate::Vec<(#krate::BenchmarkParameter, u32, u32)> {
+				#krate::vec! [
+					#(
+						(#krate::BenchmarkParameter::#param_ranges)
+					),*
+				]
+			}
+
+			fn instance(
+				&self,
+				components: &[(#krate::BenchmarkParameter, u32)],
+				verify: bool
+			) -> Result<#krate::Box<dyn FnOnce() -> Result<(), #krate::BenchmarkError>>, #krate::BenchmarkError> {
+				#(
+					// prepare instance #param_names
+					let #param_names = components.iter()
+						.find(|&c| c.0 == #krate::BenchmarkParameter::#param_names)
+						.ok_or("Could not find component during benchmark preparation.")?
+						.1;
+				)*
+
+				// benchmark setup code
+				#(
+					#setup_stmts
+				)*
+				#pre_call
+				Ok(#krate::Box::new(move || -> Result<(), #krate::BenchmarkError> {
+					#post_call
+					if verify {
+						#(
+							#verify_stmts
+						)*
+					}
+					Ok(())
+				}))
+			}
+		}
+
+		#[cfg(test)]
+		impl<#full_generics> Pallet<#generics> where T: ::frame_system::Config, #where_clause {
+			#[allow(unused)]
+			fn #test_ident() -> Result<(), #krate::BenchmarkError> {
+				let selected_benchmark = SelectedBenchmark::#name;
+				let components = <
+					SelectedBenchmark as #krate::BenchmarkingSetup<T, _>
+				>::components(&selected_benchmark);
+				let execute_benchmark = |
+					c: #krate::Vec<(#krate::BenchmarkParameter, u32)>
+				| -> Result<(), #krate::BenchmarkError> {
+					// Always reset the state after the benchmark.
+					#krate::defer!(#krate::benchmarking::wipe_db());
+
+					// Set up the benchmark, return execution + verification function.
+					let closure_to_verify = <
+						SelectedBenchmark as #krate::BenchmarkingSetup<T, _>
+					>::instance(&selected_benchmark, &c, true)?;
+
+					// Set the block number to at least 1 so events are deposited.
+					if #krate::Zero::is_zero(&frame_system::Pallet::<T>::block_number()) {
+						frame_system::Pallet::<T>::set_block_number(1u32.into());
+					}
+
+					// Run execution + verification
+					closure_to_verify()
+				};
+
+				if components.is_empty() {
+					execute_benchmark(Default::default())?;
+				} else {
+					let num_values: u32 = if let Ok(ev) = std::env::var("VALUES_PER_COMPONENT") {
```

Comment:
Should we load it into a lazy_static somewhere and reference that in both places?

Reply:
I mean the whole code here is like 300 lines of copy&paste, or?

Reply:
Essentially. The syntax of the interpolations is different but other than that its a lot of copy paste. I don't think there is a way around this though. Eventually if we deprecate the old syntax, we could way down the road eliminate the duplication, but other than that I don't think there is a way to have this in the same place and cater to the two different interpolation styles, etc. Open to any ideas though!

Reply:
Maybe we can create *another* macro that expands to this code.  
I have to change it occasionally and it would inevitably get out-of-sync.  
That would not be an ideal solution either…

Reply:
I guess it's a question of which is more annoying, maintaining both for a while or writing a whole new macro to generate it in two places before eventually deprecating the old place

Reply:
I'd be happy to explore this consolidation in a new PR soon though

Reply:
any further thoughts on this or can I mark as resolved @ggwpez?